### ISSUE TITLE
Ethernet not working on NanoPi Neo for core-image-base

### ISSUE THREAD

**Issue by TheMeaningfulEngineer** [linux-sunxi/meta-sunxi#194]
The following build steps don't result in an image with working ethernet.
Just copy pasting them should result in an image build.

```
mkdir yocto
cd yocto
export YOCTO_DIR=`pwd`
git clone git://git.openembedded.org/meta-openembedded meta-openembedded --branch rocko
git clone git://git.openembedded.org/openembedded-core oe --branch rocko
git clone git://git.openembedded.org/bitbake oe/bitbake --branch 1.36
git clone git://github.com/linux-sunxi/meta-sunxi.git meta-sunxi 
source oe/oe-init-build-env tmp
sed -i 's,MACHINE ??= .*,MACHINE= "nanopi-neo",' conf/local.conf
sed -i '5,10d' conf/bblayers.conf
echo 'BBLAYERS ?= " \'                                     >> conf/bblayers.conf
echo  $YOCTO_DIR'/oe/meta \'                             >> conf/bblayers.conf
echo  $YOCTO_DIR'/meta-openembedded/meta-oe \'           >> conf/bblayers.conf
echo  $YOCTO_DIR'/meta-openembedded/meta-networking \'   >> conf/bblayers.conf
echo  $YOCTO_DIR'/meta-openembedded/meta-python \'       >> conf/bblayers.conf
echo  $YOCTO_DIR'/meta-sunxi "'                          >> conf/bblayers.conf
bitbake core-image-base

#Flashed the core-image-base-nanopi-neo-20171208175722.rootfs.sunxi-sdimg
```

Boot log and trying to run ethernet:

```
U-Boot SPL 2017.03-g8537ddd (Dec 08 2017 - 16:37:16)
DRAM: 512 MiB
Trying to boot from MMC1


U-Boot 2017.03-g8537ddd (Dec 08 2017 - 16:37:16 +0100) Allwinner Technology

CPU:   Allwinner H3 (SUN8I 1680)
Model: FriendlyARM NanoPi NEO
DRAM:  512 MiB
MMC:   SUNXI SD/MMC: 0
*** Warning - bad CRC, using default environment

In:    serial
Out:   serial
Err:   serial
Net:   phy interface0
eth0: ethernet@1c30000
starting USB...
USB0:   USB EHCI 1.00
USB1:   USB OHCI 1.0
scanning bus 0 for devices... 1 USB Device(s) found
       scanning usb for storage devices... 0 Storage Device(s) found
Hit any key to stop autoboot:  0 
switch to partitions #0, OK
mmc0 is current device
Scanning mmc 0:1...
Found U-Boot script /boot.scr
reading /boot.scr
714 bytes read in 18 ms (38.1 KiB/s)
## Executing script at 43100000
reading sun8i-h3-nanopi-neo.dtb
13890 bytes read in 25 ms (542 KiB/s)
reading zImage
** Unable to read file zImage **
reading boot/zImage
** Unable to read file boot/zImage **
reading uImage
3853256 bytes read in 217 ms (16.9 MiB/s)
## Booting kernel from Legacy Image at 42000000 ...
   Image Name:   Linux-4.14.0-rc3
   Image Type:   ARM Linux Kernel Image (uncompressed)
   Data Size:    3853192 Bytes = 3.7 MiB
   Load Address: 40008000
   Entry Point:  40008000
   Verifying Checksum ... OK
## Flattened Device Tree blob at 43000000
   Booting using the fdt blob at 0x43000000
   Loading Kernel Image ... OK
   Loading Device Tree to 49ff9000, end 49fff641 ... OK

Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 4.14.0-rc3 (oe-user@oe-host) (gcc version 7.2.0 (GCC)) #1 SMP Fri Dec 8 16:33:34 CET 2017
[    0.000000] CPU: ARMv7 Processor [410fc075] revision 5 (ARMv7), cr=10c5387d
[    0.000000] CPU: div instructions available: patching division code
[    0.000000] CPU: PIPT / VIPT nonaliasing data cache, VIPT aliasing instruction cache
[    0.000000] OF: fdt: Machine model: FriendlyARM NanoPi NEO
[    0.000000] Memory policy: Data cache writealloc
[    0.000000] psci: probing for conduit method from DT.
[    0.000000] psci: Using PSCI v0.1 Function IDs from DT
[    0.000000] random: fast init done
[    0.000000] percpu: Embedded 16 pages/cpu @dffad000 s34124 r8192 d23220 u65536
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 130048
[    0.000000] Kernel command line: console=ttyS0,115200 console=tty1 root=/dev/mmcblk0p2 rootwait panic=10
[    0.000000] PID hash table entries: 2048 (order: 1, 8192 bytes)
[    0.000000] Dentry cache hash table entries: 65536 (order: 6, 262144 bytes)
[    0.000000] Inode-cache hash table entries: 32768 (order: 5, 131072 bytes)
[    0.000000] Memory: 509508K/524288K available (6144K kernel code, 412K rwdata, 1316K rodata, 1024K init, 266K bss, 14780K reserved, 0K cma-reserved, 0K highmem)
[    0.000000] Virtual kernel memory layout:
[    0.000000]     vector  : 0xffff0000 - 0xffff1000   (   4 kB)
[    0.000000]     fixmap  : 0xffc00000 - 0xfff00000   (3072 kB)
[    0.000000]     vmalloc : 0xe0800000 - 0xff800000   ( 496 MB)
[    0.000000]     lowmem  : 0xc0000000 - 0xe0000000   ( 512 MB)
[    0.000000]     pkmap   : 0xbfe00000 - 0xc0000000   (   2 MB)
[    0.000000]     modules : 0xbf000000 - 0xbfe00000   (  14 MB)
[    0.000000]       .text : 0xc0008000 - 0xc0700000   (7136 kB)
[    0.000000]       .init : 0xc0900000 - 0xc0a00000   (1024 kB)
[    0.000000]       .data : 0xc0a00000 - 0xc0a672c0   ( 413 kB)
[    0.000000]        .bss : 0xc0a6e2bc - 0xc0ab0e34   ( 267 kB)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] Hierarchical RCU implementation.
[    0.000000] 	RCU restricting CPUs from NR_CPUS=8 to nr_cpu_ids=4.
[    0.000000] RCU: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=4
[    0.000000] NR_IRQS: 16, nr_irqs: 16, preallocated irqs: 16
[    0.000000] GIC: Using split EOI/Deactivate mode
[    0.000000] clocksource: timer: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 79635851949 ns
[    0.000000] arch_timer: cp15 timer(s) running at 24.00MHz (phys).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0x588fe9dc0, max_idle_ns: 440795202592 ns
[    0.000006] sched_clock: 56 bits at 24MHz, resolution 41ns, wraps every 4398046511097ns
[    0.000017] Switching to timer-based delay loop, resolution 41ns
[    0.000192] Console: colour dummy device 80x30
[    0.000534] console [tty1] enabled
[    0.000568] Calibrating delay loop (skipped), value calculated using timer frequency.. 48.00 BogoMIPS (lpj=240000)
[    0.000593] pid_max: default: 32768 minimum: 301
[    0.000728] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.000749] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.001375] CPU: Testing write buffer coherency: ok
[    0.001749] /cpus/cpu@0 missing clock-frequency property
[    0.001779] /cpus/cpu@1 missing clock-frequency property
[    0.001802] /cpus/cpu@2 missing clock-frequency property
[    0.001824] /cpus/cpu@3 missing clock-frequency property
[    0.001843] CPU0: thread -1, cpu 0, socket 0, mpidr 80000000
[    0.002189] Setting up static identity map for 0x40100000 - 0x40100060
[    0.002323] Hierarchical SRCU implementation.
[    0.002821] smp: Bringing up secondary CPUs ...
[    0.013479] CPU1: thread -1, cpu 1, socket 0, mpidr 80000001
[    0.024210] CPU2: thread -1, cpu 2, socket 0, mpidr 80000002
[    0.034907] CPU3: thread -1, cpu 3, socket 0, mpidr 80000003
[    0.034981] smp: Brought up 1 node, 4 CPUs
[    0.035037] SMP: Total of 4 processors activated (192.00 BogoMIPS).
[    0.035049] CPU: All CPU(s) started in HYP mode.
[    0.035060] CPU: Virtualization extensions available.
[    0.035839] devtmpfs: initialized
[    0.039022] VFP support v0.3: implementor 41 architecture 2 part 30 variant 7 rev 5
[    0.039269] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.039304] futex hash table entries: 1024 (order: 4, 65536 bytes)
[    0.039564] pinctrl core: initialized pinctrl subsystem
[    0.040290] NET: Registered protocol family 16
[    0.040649] DMA: preallocated 256 KiB pool for atomic coherent allocations
[    0.041466] cpuidle: using governor menu
[    0.041655] hw-breakpoint: found 5 (+1 reserved) breakpoint and 4 watchpoint registers.
[    0.041678] hw-breakpoint: maximum watchpoint size is 8 bytes.
[    0.050881] SCSI subsystem initialized
[    0.051277] usbcore: registered new interface driver usbfs
[    0.051332] usbcore: registered new interface driver hub
[    0.051393] usbcore: registered new device driver usb
[    0.051495] pps_core: LinuxPPS API ver. 1 registered
[    0.051510] pps_core: Software ver. 5.3.6 - Copyright 2005-2007 Rodolfo Giometti <giometti@linux.it>
[    0.051540] PTP clock support registered
[    0.052374] clocksource: Switched to clocksource arch_sys_counter
[    0.059105] NET: Registered protocol family 2
[    0.059727] TCP established hash table entries: 4096 (order: 2, 16384 bytes)
[    0.059793] TCP bind hash table entries: 4096 (order: 3, 32768 bytes)
[    0.059864] TCP: Hash tables configured (established 4096 bind 4096)
[    0.059964] UDP hash table entries: 256 (order: 1, 8192 bytes)
[    0.060015] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes)
[    0.060199] NET: Registered protocol family 1
[    0.060594] RPC: Registered named UNIX socket transport module.
[    0.060616] RPC: Registered udp transport module.
[    0.060628] RPC: Registered tcp transport module.
[    0.060639] RPC: Registered tcp NFSv4.1 backchannel transport module.
[    0.062802] workingset: timestamp_bits=30 max_order=17 bucket_order=0
[    0.067785] NFS: Registering the id_resolver key type
[    0.067849] Key type id_resolver registered
[    0.067861] Key type id_legacy registered
[    0.069179] Block layer SCSI generic (bsg) driver version 0.4 loaded (major 249)
[    0.069211] io scheduler noop registered
[    0.069223] io scheduler deadline registered
[    0.069443] io scheduler cfq registered (default)
[    0.069459] io scheduler mq-deadline registered
[    0.069472] io scheduler kyber registered
[    0.073389] sun8i-h3-pinctrl 1c20800.pinctrl: initialized sunXi PIO driver
[    0.074719] sun8i-h3-r-pinctrl 1f02c00.pinctrl: initialized sunXi PIO driver
[    0.118766] Serial: 8250/16550 driver, 8 ports, IRQ sharing disabled
[    0.120967] console [ttyS0] disabled
[    0.141217] 1c28000.serial: ttyS0 at MMIO 0x1c28000 (irq = 30, base_baud = 1500000) is a U6_16550A
[    0.762179] console [ttyS0] enabled
[    0.767073] libphy: Fixed MDIO Bus: probed
[    0.771421] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    0.777990] ehci-platform: EHCI generic platform driver
[    0.783435] ehci-platform 1c1d000.usb: EHCI Host Controller
[    0.789036] ehci-platform 1c1d000.usb: new USB bus registered, assigned bus number 1
[    0.796940] ehci-platform 1c1d000.usb: irq 24, io mem 0x01c1d000
[    0.832395] ehci-platform 1c1d000.usb: USB 2.0 started, EHCI 1.00
[    0.839183] hub 1-0:1.0: USB hub found
[    0.843017] hub 1-0:1.0: 1 port detected
[    0.847342] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    0.853542] ohci-platform: OHCI generic platform driver
[    0.858936] ohci-platform 1c1d400.usb: Generic Platform OHCI controller
[    0.865601] ohci-platform 1c1d400.usb: new USB bus registered, assigned bus number 2
[    0.873428] ohci-platform 1c1d400.usb: irq 25, io mem 0x01c1d400
[    0.946959] hub 2-0:1.0: USB hub found
[    0.950753] hub 2-0:1.0: 1 port detected
[    0.955391] usbcore: registered new interface driver usb-storage
[    0.961765] sun6i-rtc 1f00000.rtc: rtc core: registered rtc-sun6i as rtc0
[    0.968582] sun6i-rtc 1f00000.rtc: RTC enabled
[    0.973165] i2c /dev entries driver
[    0.976853] IR NEC protocol handler initialized
[    0.981386] IR RC5(x/sz) protocol handler initialized
[    0.986449] IR RC6 protocol handler initialized
[    0.990980] IR JVC protocol handler initialized
[    0.995519] IR Sony protocol handler initialized
[    1.000134] IR SANYO protocol handler initialized
[    1.004844] IR Sharp protocol handler initialized
[    1.009547] IR MCE Keyboard/mouse protocol handler initialized
[    1.015389] IR XMP protocol handler initialized
[    1.020257] sunxi-wdt 1c20ca0.watchdog: Watchdog enabled (timeout=16 sec, nowayout=0)
[    1.028402] cpu cpu0: failed to get clock: -2
[    1.032800] cpufreq-dt: probe of cpufreq-dt failed with error -2
[    1.039625] sunxi-mmc 1c0f000.mmc: Got CD GPIO
[    1.102398] sunxi-mmc 1c0f000.mmc: base:0xe0874000 irq:23
[    1.108497] ledtrig-cpu: registered to indicate activity on CPUs
[    1.115028] usbcore: registered new interface driver usbhid
[    1.120614] usbhid: USB HID core driver
[    1.125139] NET: Registered protocol family 10
[    1.130600] Segment Routing with IPv6
[    1.134392] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    1.140872] NET: Registered protocol family 17
[    1.145444] Key type dns_resolver registered
[    1.149825] Registering SWP/SWPB emulation handler
[    1.158896] registered taskstats version 1
[    1.163175] sun6i-rtc 1f00000.rtc: setting system clock to 1970-01-01 00:00:06 UTC (6)
[    1.171345] vcc3v0: disabling
[    1.174339] vcc5v0: disabling
[    1.177502] Waiting for root device /dev/mmcblk0p2...
[    1.199402] mmc0: host does not support reading read-only switch, assuming write-enable
[    1.209415] mmc0: new high speed SDHC card at address 59b4
[    1.215318] mmcblk0: mmc0:59b4 USD   7.51 GiB 
[    1.220962]  mmcblk0: p1 p2
[    1.253169] EXT4-fs (mmcblk0p2): couldn't mount as ext3 due to feature incompatibilities
[    1.261777] EXT4-fs (mmcblk0p2): couldn't mount as ext2 due to feature incompatibilities
[    1.270742] EXT4-fs (mmcblk0p2): INFO: recovery required on readonly filesystem
[    1.278073] EXT4-fs (mmcblk0p2): write access will be enabled during recovery
[    1.289017] EXT4-fs (mmcblk0p2): recovery complete
[    1.296584] EXT4-fs (mmcblk0p2): mounted filesystem with ordered data mode. Opts: (null)
[    1.304757] VFS: Mounted root (ext4 filesystem) readonly on device 179:2.
[    1.315047] devtmpfs: mounted
[    1.319140] Freeing unused kernel memory: 1024K
[    1.616184] udevd[104]: starting version 3.2.2
[    1.689929] udevd[105]: starting eudev-3.2.2
[    2.132247] EXT4-fs (mmcblk0p2): re-mounted. Opts: data=ordered
[    2.899368] Bluetooth: Core ver 2.22
[    2.903173] NET: Registered protocol family 31
[    2.907664] Bluetooth: HCI device and connection manager initialized
[    2.914135] Bluetooth: HCI socket layer initialized
[    2.919067] Bluetooth: L2CAP socket layer initialized
[    2.924193] Bluetooth: SCO socket layer initialized
[    3.026464] Bluetooth: BNEP (Ethernet Emulation) ver 1.3
[    3.031826] Bluetooth: BNEP filters: protocol 
[    3.036378] Bluetooth: BNEP socket layer initialized

OpenEmbedded nodistro.0 nanopi-neo /dev/ttyS0

nanopi-neo login: root
root@nanopi-neo:~# ifconfig -a
lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1%1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

sit0      Link encap:IPv6-in-IPv4  
          NOARP  MTU:1480  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

root@nanopi-neo:~# 
```

**Comment by NguyenTrongThinh** [comment]
In Kernel 4.14, emac driver is unavailable. see this link http://linux-sunxi.org/Sun8i_emac

**Comment by TheMeaningfulEngineer** [comment]
From http://linux-sunxi.org/Sun8i_emac:
> The second is dwmac-sun8i, this a glue to the stmmac driver. You can find the latest version at https://github.com/montjoie/linux/tree/dwmac-sun8i-current
This driver is mainline, but DT was reverted in 4.13-rc7. DT should be back soon.

The DT (Device Tree) was reverted not to support the mainlined driver?
Am I getting this right (because it seems somewhat strange)?

So it's just a matter of a correct device tree with the mainline 3.14 to get it running again?

**Comment by NguyenTrongThinh** [comment]
I use the first driver which is discontinued. Here is source kernel  https://github.com/montjoie/linux/tree/sun8i-emac-wip

I add these to device tree

**file sun8i-h3-nanopi.dtsi**

```
+&emac {
+	phy = <&phy1>;
+	phy-mode = "mii";
+	allwinner,use-internal-phy;
+	allwinner,leds-active-low;
+	status = "okay";
+	phy1: ethernet-phy@1 {
+		reg = <1>;
+	};
+};
```
**file sun8i-h3.dtsi**
```
+		emac: ethernet@1c30000 {
+			compatible = "allwinner,sun8i-h3-emac";
+			reg = <0x01c30000 0x104>, <0x01c00030 0x4>;
+			reg-names = "emac", "syscon";
+			interrupts = <GIC_SPI 82 IRQ_TYPE_LEVEL_HIGH>;
+			resets = <&ccu RST_BUS_EMAC>, <&ccu RST_BUS_EPHY>;
+			reset-names = "ahb", "ephy";
+			clocks = <&ccu CLK_BUS_EMAC>, <&ccu CLK_BUS_EPHY>;
+			clock-names = "ahb", "ephy";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
```

**Comment by MicroJoe** [comment]
> So it's just a matter of a correct device tree with the mainline 3.14 to get it running again?

I hope you meant 4.14 here. I am using 4.14 from linux-megous [1] that provides a patched device-tree for Nanopi NEO dwmac-sun8i support [2] as it was not present in mainline 4.14 [3]. It is however available in 4.15 release [4].

However I am in the process of migration from linux-megous to linux-mainline with this meta-sunxi layer. I guess I can propose a patch for the device tree for stable 4.13 kernel used here.

[1] : https://github.com/megous/linux
[2] : https://github.com/megous/linux/blob/orange-pi-4.14/arch/arm/boot/dts/sun8i-h3-nanopi-neo.dts
[3] : http://elixir.free-electrons.com/linux/v4.14.16/source/arch/arm/boot/dts/sun8i-h3-nanopi-neo.dts
[4] : http://elixir.free-electrons.com/linux/v4.15/source/arch/arm/boot/dts/sun8i-h3-nanopi-neo.dts

**Comment by ebutera** [comment]
linux-mainline is now at 4.16.13 so it should be fixed, please reopen if it is not.
